src/server/dfly_main.cc

```diff
@@ -557,68 +581,93 @@ void UpdateResourceLimitsIfInsideContainer(io::MemInfoData* mdata, size_t* max_t
     return;
   }
 
-  auto original_memory = mdata->mem_total;
+  CHECK(!err) << "Unsupported cgroup v2 format, read:" << mem_path;
+
+  VLOG(1) << "mem_path = " << mem_path;
+  VLOG(1) << "cpu_path = " << cpu_path;
 
   /* Update memory limits */
-  auto mlimit = io::ReadFileToString(mem_path + "/memory.limit_in_bytes");
-  DVLOG(1) << "memory/memory.limit_in_bytes: " << mlimit.value_or("N/A");
 
-  if (mlimit && !absl::StartsWith(*mlimit, "max")) {
-    CHECK(absl::SimpleAtoi(*mlimit, &mdata->mem_total));
-  }
+  // Start by reading global memory limits
+  constexpr auto base_mem = "/sys/fs/cgroup/memory"sv;
+  read_mem(StrCat(base_mem, "/memory.limit_in_bytes"), &mdata->mem_total);
+  read_mem(StrCat(base_mem, "/memory.max"), &mdata->mem_total);
 
-  auto mmax = io::ReadFileToString(mem_path + "/memory.max");
-  DVLOG(1) << "memory.max: " << mmax.value_or("N/A");
+  // Read cgroup-specific limits
+  read_mem(StrCat(mem_path, "/memory.limit_in_bytes"), &mdata->mem_total);
+  read_mem(StrCat(mem_path, "/memory.max"), &mdata->mem_total);
+  read_mem(StrCat(mem_path, "/memory.high"), &mdata->mem_avail);
 
-  if (mmax && !absl::StartsWith(*mmax, "max")) {
-    CHECK(absl::SimpleAtoi(*mmax, &mdata->mem_total));
-  }
+  mdata->mem_avail = min(mdata->mem_avail, mdata->mem_total);
```

Comment:
Can the total memory ever be less than the available memory?

Reply:
oh good catch! because `read_mem()` also updates via `min` I think this `min` is actually redundant!

Reply:
edit: wait no that is not right. If we set `memory.max` to `8G`, not set `memory.high`, and the memory has more than 8G of available memory then the assignment does `mem_avail = mem_total`.

On the other hand, if the available memory is less than what was set in `mem_total`, the assignment is `mem_avail = mem_total`.

So I don't think its possible to remove the `min()` here.